index/index.go

```diff
@@ -0,0 +1,498 @@
+		}
+		if err = i.putSealedID(lastLabelUUID, &lastSealedID, true); err != nil {
+			return err
+		}
+	}
+
+	// Compute new labelUUID and plaintext ID, seal it, and send it to the IO Provider.
+	labelUUID, err := computeLabelUUID(counter, &tagger)
+	if err != nil {
+		return err
+	}
+
+	plainID := PlainID{DocID: docID, NextCounter: 0}
+
+	sealedID, err := plainID.Seal(labelUUID, &cryptor)
+	if err != nil {
+		return err
+	}
+	if err = i.putSealedID(labelUUID, &sealedID, false); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Delete deletes a keyword/docID pair from the secure index.
+func (i *SecureIndex) Delete(key []byte, token, keyword, docID string) error {
+	identity, err := i.idProvider.GetIdentity(token)
+	if err != nil {
+		return ErrNotAuthenticated
+	}
+	if !identity.Scopes.Contains(id.ScopeIndex) {
+		return ErrNotAuthorized
+	}
+
+	if len(key) != MasterKeyLength {
+		return ErrInvalidMasterKeyLength
+	}
+
+	// Create two keys, k1 and k2: One for tagger and one for cryptor.
+	k1 := crypto.KMACKDF(crypto.TaggerKeyLength, key, []byte("label"), []byte(keyword))
+	k2 := crypto.KMACKDF(crypto.EncryptionKeyLength, key, []byte("id"), []byte(keyword))
+
+	tagger, err := crypto.NewKMAC256Tagger(k1)
+	if err != nil {
+		return err
+	}
+	cryptor, err := crypto.NewAESCryptor(k2)
+	if err != nil {
+		return err
+	}
+
+	counter := uint64(0)
+
+	previousSealedID := SealedID{}
+	previousLabel := uuid.Nil
+
+	// Starting with counter = 0, compute the corresponding sealed ID, unseal it,
+	// and check if its DocID is equal to the docID given as input. Otherwise,
+	// update counter to the next counter and repeat.
+	for {
+		labelUUID, err := computeLabelUUID(counter, &tagger)
+		if err != nil {
+			return err
+		}
+
+		// Get the sealed ID from the IO Provider.
+		sealedID, err := i.getSealedID(labelUUID)
+		if err != nil {
+			return err
+		}
+
+		decryptedID, err := sealedID.Unseal(labelUUID, &cryptor)
+		if err != nil {
+			return err
+		}
+
+		// Check if the current DocID matches the docID given as input.
+		if decryptedID.DocID == docID {
+			// Delete the sealed ID.
+			if err = i.deleteSealedID(labelUUID); err != nil {
+				return err
+			}
+
+			if previousSealedID.Ciphertext != nil {
+				// In this case, there exists a previous sealed ID which should be updated:
+				// The previous sealed ID's next counter should be updated to the deleted sealed ID's next counter.
+				return i.updatePreviousAfterDelete(previousLabel, previousSealedID, decryptedID.NextCounter, &cryptor)
+			} else if decryptedID.NextCounter != 0 {
+				// In this case, there exists a next sealed ID which should be updated:
+				// The next sealed ID's counter should be updated to 0 as this sealed ID is now the first.
+				return i.updateNextAfterDelete(decryptedID.NextCounter, labelUUID, &cryptor, &tagger)
+			}
+			// Else, the sealed ID that was deleted was the only sealed ID containing the given keyword,
+			// hence nothing else has to be done.
+			return nil
```

Comment:
What happens if someone adds the keyword/id pair more than once? I think this only deletes the first one?